### DISCUSSION TITLE
How to group or sort in a query?

### DISCUSSION THREAD

**Discussion by vkensou** [friflo/Friflo.Engine.ECS#25] [Q&A] [answered]
I want to batch process objects with the same material and grid. How should I do it?

```C#
public struct ECSMesh : IComponent
{
    public Mesh mesh;
}

public struct ECSMaterial : IComponent
{
    public Material material;
}

public class RenderSystem : QuerySystem<ECSMesh, ECSMaterial, Position>
{
    protected override void OnUpdate()
    {
// batch process same mesh and same material
    }
}

```

**Comment by friflo** [comment, marked as answer]
The ECS has currently not functionality for sorting.
You can use `query.ToEntityList().ToArray()` to get the entities of a query as an `Entity[]`.
This array can than be sorted with `Array.Sort<>()`.

See [Example](https://github.com/friflo/Friflo.Engine.ECS/blob/main/src/Tests/ECS/Github/Test_GitHub_25.cs)

**Reply by vkensou** [reply]
> The ECS has currently not functionality for sorting. You can use `query.ToEntityList().ToArray()` to get the entities of a query as an `Entity[]`. This array can than be sorted with `Array.Sort<>()`.
> 
> See [Example](https://github.com/friflo/Friflo.Engine.ECS/blob/main/src/Tests/ECS/Github/Test_GitHub_25.cs)

Thank you reply. Seems I must copy datas from ecs to outside world.

**Reply by friflo** [reply]
> Seems I must copy datas from ecs to outside world.

Correct.

Also, an optimization could be storing Entity & Component in a single struct and use this in an array for sorting.
This would minimize the `.GetComponent<MyComponent1>()` calls.
